README.md
# ckgdoku
Version of ckgedit which allows for switching between native dokuwiki editor tools and ckgedit's tools.  

Installed in its own directory and treated as separate from ckgedit. However, the ckgedit documentation
on Dokuwiki.org applies to ckgedoku as well.

Ckgdoku and ckgedit cannot both be enabled at the same time.  


Changes
rev_18-May_21-23_21

Switched dafault filebrowser settings to all and ckgedit

18 Nov
Fix for HTML in user profile panel

2 Feb 2017
Synched with ckgedit

3 March 
remove backslashes from ends of blockquote

Mar 21
Converted all uses of ckgedit to ckgdoku in code; now it loads as plugin ckgdoku

Mar 22
ckgedit/ckgdoku warning: one must be disabled (meta.php)

Mar 23
updated localization of symlink creation in meta.php

Mar 24
Merged trans into master

Mar 28
Fix in script.js for assigning image parameters (size, link, align)

Mar 29
Synced to captcha

Mar 30
Fixed hiding specialized buttons for non-users wehn captcha is on

Apr 1
removed extra / from fileurl sent from media manager to ckeditor link dialog in script.js (pull  #9) 

Apr 4
Moved javascript from mediamanager.php to scripts/mediamgr.js, loaded by TPL_METAHEADER_OUTPUT.

Apr 5
support for Hover texts on images

Apr 6
Fixes missing curly at end of mediamgr.js

Apr 20
New setting htmbloc_okk gives option of allowing HTML_BLOCKs without setti…ng DW's htmlok to true

Apr 22
Double click edit section opens dw editor

Apr 23
Implemented reminders of new double-click feature for Admins

May 2
Fixed bug in meta.php htmblock_ok -> htmlblock_ok for getting config option

May 4
Added Geshi plugin

May 5
refactored geshi.js->getHref()

May 6
tweaked geshi icon
Fixed JS type error for mis-labled fr and de lang files
Geshi German language file contributed by github user

May 7
Placed geshi progamming lang selection into variable (had been hard-coded to java for code blocks)

May 11
Fixed wrong URL to ajax.php when loading GeSHi options (on Nginx)

May 24
Fixed broken list when li contains links
 
June 1
Enabled code and file blolcks and snippets in tables
Fixed cell where snippet together with text breaks cell

June 2
Replaced faulty skayt with skayt 4.5.7
Replaced moono with 4.7 moono
Removed Samples
Fixed references in ckedit.js that still pointed to ckgedit
5
Jan 3
Restored orignal link.js for ckgdoku, containing linkwiz, which had been inadvertently exchanged for ckgedit's
Fixed deprecated synchrononous post in link.js

Jun4
CKEDitor 4.7 ready to go
Switching to tag 'current' as Dokuwiki download version
Added dblclk option

Jun 5
Updated Javascript for new dblick option

Jun 12/13
Support for canonical urls
automate config for CKEditor extra plugins
Automate addition of xtra plugins to xtra plugins list

Jan 14/Jan16
Fixed canonical url/path in config.js

June 21
Fixed error in meta.php when accessesing none-existent wrap->getConf

Jun 23
Added CKEditor pastbase64 plugin for copy and paste

Jul 27
Added Internal Link Texxt Option to Link Dialog

Aug 3
Fixed multiple urlencoding of interwiki links

Aug 4
Merged ilinkt branch: creates correct root ns internal links when using manual entries in link tool

Aug 23
Support for switching between CKEdiktor GUIs

Sep 7
switches filling interwiki select array from setTimeout to set Interval
smoother switching betweein image paste modes

Sep 18
Added initial colon to links returned from linkwiz, to establish them as root, otherwise they are configured as
belonging to current namespace

Sep 24 
jQuery support for edittable plugin

Oct 22
Insure namespace template data is saved when there has been no activity in editor

Oct 24
Fixes broken table when cell has pipe in nowiki

Oct 26
Adds back in backslashes when code/file blocks are filtered for extra final line-feeds in tables

Nov 6
Fixed bug in link.js where ckgdokuIwikiIndex was left in 2 places as ckgeditIwikiIndex

Nov 28
Merged farms branch into master

Dec 5
Fix iwiki handling to process more than one type extension type.12.12, etc

Dec 8
merge iwiki branch into master
in save.php fix deleted backslash in code blocks

Dec 19
fixed https://github.com/turnermm/ckgedit/issues/360: html tag markup confused with interwiki markup: >

Dec 21
Fixes https://github.com/turnermm/ckgdoku/issues/45 (left-over interwiki internal markup removed)

Dec  27
Fixed interwiki urls that start with colon and get changed to wikilinks

Dec 30
Fixes ckgdoku/issues/47: mishandled image with relative links

Dec 31
Fixes https://github.com/turnermm/ckgdoku/issues/48, page duplication on switching to dwedit
same as https://github.com/turnermm/ckgedit/issues/343

Jan 2 2018
Update fix to removed backslash in code blocks

Jan 4
Copied some German language updates from ckgedit

Jan 8
Fix for backslashes removed on save: https://github.com/turnermm/ckgdoku/issues/27

Jan 10
 Removed oIWIKIo inserted after '>' when that is followed by link text

Jan 11
Update fix for turnermm/ckgdoku/issues/27
fixed typo in edit.php
fix faulty regex to restore valid backslashes when removing extra line-breaks from code blocks

Jan 16
fix corrupted image as link when checking for userheading and relative links

Jan 31
Restores query string to internal link, missing after coded for relative links

Apr 17
Adds missing HTML_BLOCK aliases


script.js
/*
 * HTML Parser By John Resig (ejohn.org)
 * Original code by Erik Arvidsson, Mozilla Public License
 * http://erik.eae.net/simplehtmlparser/simplehtmlparser.js
 * @license    GPL 3 or later (http://www.gnu.org/licenses/gpl.html)
*/

var HTMLParser;
var HTMLParserInstalled=true;
var HTMLParser_Elements = new Array(); 
(function(){

	// Regular Expressions for parsing tags and attributes
	var startTag = /^<(\w+)((?:\s+\w+(?:\s*=\s*(?:(?:"[^"]*")|(?:'[^']*')|[^>\s]+))?)*)\s*(\/?)>/,
		endTag = /^<\/(\w+)[^>]*>/,
		attr = /(\w+)(?:\s*=\s*(?:(?:"((?:\\.|[^"])*)")|(?:'((?:\\.|[^'])*)')|([^>\s]+)))?/g;
		
	// Empty Elements - HTML 4.01
	var empty = makeMap("br,col,hr,img");
   // HTMLParser_Elements['empty'] = empty;

	// Block Elements - HTML 4.01
	var block = makeMap("blockquote,center,del,div,dl,dt,hr,iframe,ins,li,ol,p,pre,table,tbody,td,tfoot,th,thead,tr,ul");
  //  HTMLParser_Elements['block'] = block;

	// Inline Elements - HTML 4.01
	var inline = makeMap("a,abbr,acronym,b,big,br,cite,code,del,em,font,h1,h2,h3,h4,h5,h6,i,img,ins,kbd,q,s,samp,small,span,strike,strong,sub,sup,tt,u,var");

	// Elements that you can, intentionally, leave open
	// (and which close themselves)
	var closeSelf = makeMap("colgroup,dd,dt,li,options,p,td,tfoot,th,thead,tr");

	// Attributes that have their values filled in disabled="disabled"
	var fillAttrs = makeMap("checked,disabled,ismap,noresize,nowrap,readonly,selected");

	// Special Elements (can contain anything)
	var special = makeMap("script,style");
 
   
	HTMLParser = this.HTMLParser = function( html, handler ) {
		var index, chars, match, stack = [], last = html;      
          html = html.replace(/~~OPEN_HTML_BLOCK~~/gm , '~~START_HTML_BLOCK~~') ;
          html = html.replace(/~~END_HTML_BLOCK~~/gm , '~~CLOSE_HTML_BLOCK~~') ;
         if(html.match(/~~START_HTML_BLOCK~~/gm) ){            //adopted [\s\S] from Goyvaerts, Reg. Exp. Cookbook (O'Reilly)
              if(!JSINFO['htmlok']) {
                 html = html.replace(/~~START_HTML_BLOCK~~|~~CLOSE_HTML_BLOCK~~/gm,"");
                } 
        
             html = html.replace(/(<p.*?>)*\s*~~START_HTML_BLOCK~~\s*(<\/p>)*([\s\S]+)~~CLOSE_HTML_BLOCK~~\s*(<\/p>)*/gm, function(match,p,p1,text,p2) {       
             text = text.replace(/<\/?div.*?>/gm,"");
             text = text.replace(/<code>/gm,"");             
             text = text.replace(/<\/code>/gm,"");             
             text = text.replace(/</gm,"&lt;");
             text = text.replace(/<\//gm,"&gt;");
             return  "~~START_HTML_BLOCK~~\n\n" +   text  + "\n\n~~CLOSE_HTML_BLOCK~~\n\n";
         }); 
        }
        /* remove dwfck note superscripts from inside links */
        html = html.replace(/(<sup\s+class=\"dwfcknote fckgL\d+\"\>fckgL\d+\s*\<\/sup\>)\<\/a\>/gm, function(match,sup,a) {
             return( '</a>' +sup);   
         }
       );
        
		stack.last = function(){
			return this[ this.length - 1 ];
		};

		while ( html ) {
			chars = true;

			// Make sure we're not in a script or style element
			if ( !stack.last() || !special[ stack.last() ] ) {

				// Comment
				if ( html.indexOf("<!--") == 0 ) {
					index = html.indexOf("-->");
	
					if ( index >= 0 ) {
						if ( handler.comment )
							handler.comment( html.substring( 4, index ) );
						html = html.substring( index + 3 );
						chars = false;
					}
	
				// end tag
				} else if ( html.indexOf("</") == 0 ) {
					match = html.match( endTag );
	
					if ( match ) {
						html = html.substring( match[0].length );
						match[0].replace( endTag, parseEndTag );
						chars = false;
					}
	
				// start tag
				} else if ( html.indexOf("<") == 0 ) {
					match = html.match( startTag );
	
					if ( match ) {
						html = html.substring( match[0].length );
						match[0].replace( startTag, parseStartTag );
						chars = false;
					}
				}

				if ( chars ) {
					index = html.indexOf("<");
					
					var text = index < 0 ? html : html.substring( 0, index );
					html = index < 0 ? "" : html.substring( index );
					
					if ( handler.chars )
						handler.chars( text );
				}

			} else {
				html = html.replace(new RegExp("(.*)<\/" + stack.last() + "[^>]*>"), function(all, text){
					text = text.replace(/<!--(.*?)-->/g, "$1")
						.replace(/<!\[CDATA\[(.*?)]]>/g, "$1");

					if ( handler.chars )
						handler.chars( text );

					return "";
				});

				parseEndTag( "", stack.last() );
			}

			if ( html == last )
				throw "Parse Error: " + html;
			last = html;
		}
		
		// Clean up any remaining tags
		parseEndTag();

		function parseStartTag( tag, tagName, rest, unary ) {
			if ( block[ tagName ] ) {
				while ( stack.last() && inline[ stack.last() ] ) {
					parseEndTag( "", stack.last() );
				}
			}

			if ( closeSelf[ tagName ] && stack.last() == tagName ) {
				parseEndTag( "", tagName );
			}

			unary = empty[ tagName ] || !!unary;

			if ( !unary )
				stack.push( tagName );
			
			if ( handler.start ) {
				var attrs = [];
	
				rest.replace(attr, function(match, name) {
					var value = arguments[2] ? arguments[2] :
						arguments[3] ? arguments[3] :
						arguments[4] ? arguments[4] :
						fillAttrs[name] ? name : "";
					
					attrs.push({
						name: name,
						value: value,
						escaped: value.replace(/(^|[^\\])"/g, '$1\\\"') //"
					});
				});
	
				if ( handler.start )
					handler.start( tagName, attrs, unary );
			}
		}

		function parseEndTag( tag, tagName ) {
			// If no tag name is provided, clean shop
			if ( !tagName )
				var pos = 0;
				
			// Find the closest opened tag of the same type
			else
				for ( var pos = stack.length - 1; pos >= 0; pos-- )
					if ( stack[ pos ] == tagName )
						break;
			
			if ( pos >= 0 ) {
				// Close all the open elements, up the stack
				for ( var i = stack.length - 1; i >= pos; i-- )
					if ( handler.end )
						handler.end( stack[ i ] );
				
				// Remove the open elements from the stack
				stack.length = pos;
			}
		}
	};
	

	function makeMap(str){
		var obj = {}, items = str.split(",");
		for ( var i = 0; i < items.length; i++ )
			obj[ items[i] ] = true;
		return obj;
	}
})();


function HTMLParser_test_result(results) {

var test_str = "";
for ( i=0; i < results.length; i++) {
   var character = results.charAt(i);
   if(results.charCodeAt(i) == 10)
         character ='\\n';
   if(results.charCodeAt(i) == 32)  
         character ='SP';
   var entry =  character + ' ';
    
  test_str += entry;  
  if(results.charCodeAt(i) == 10) {
       test_str += "\n";
   }
}

if(!confirm(test_str)) return false;
return true;

}

function hide_backup_msg() {
  document.getElementById("backup_msg").style.display="none";
  return false;
}

function show_backup_msg(msg) {
  document.getElementById("backup_msg").style.display="block";
  document.getElementById("backup_msg_area").innerHTML = "Backed up to: " + msg;
  
  return false;
}

  // legacy functions 
 function remove_draft(){
 }

function dwedit_draft_delete() {
}
  // legacy functions  end
  
  function setEdHeight(h) {  
        h = parseInt(h);        
        document.cookie = 'ckgEdht=' + h +';expires="";path=' +JSINFO['doku_base'];
   }

 
  function ckgd_setImgPaste(which) {        
      var state = JSINFO['ckgEdPaste'] ? JSINFO['ckgEdPaste']  : "";
      if(state == 'on')  {
            which = 'off'  
      }
      else which = 'on';      
      JSINFO['ckgEdPaste'] = which;   
       document.cookie = 'ckgEdPaste=' + which +';expires="Thu, 18 Dec 2575 12:00:00 UTC";path=' +JSINFO['doku_base'];      
      alert(LANG.plugins.ckgdoku.ckg_paste_restart + ' ' + LANG.plugins.ckgdoku[which]);    
   }

  function GetE(e) {
       return  document.getElementById(e);
  }
var dokuBase = location.host + DOKU_BASE;

 if(window.getSelection != undefined) {   
    var doku_ckg_getSelection = window.getSelection;
    window.getSelection = function(ta) {
        if(!ta) ta = GetE("wiki__text");       
        return doku_ckg_getSelection(ta);
    };
 }
 
 function ckgdoku_seteditor_priority(m,client,dw_val_obj) {
       var which = {'Y': 'Dokuwiki', 'N': 'CKEditor'};

       if (typeof m === "undefined") {  // Safari
               if(dw_val_obj[0].checked) {
                   m= dw_val_obj[0].value;
               }
              else if(dw_val_obj[1].checked) {
                           m = dw_val_obj[1].value;
              }
       }
        var params = "dw_val=" +  m;   params += '&call=cked_selector';    params += "&dwp_client=" + client;
        jQuery.post( DOKU_BASE + 'lib/exe/ajax.php', params,
                function (data) {  
                    if(data == 'done') { 
                        if(!m)
                             alert(LANG.plugins.ckgdoku.dwp_not_sel); 
                          else 
                             alert(LANG.plugins.ckgdoku.dwp_updated + which[m]); 
                    }
                      else  {
                          alert(LANG.plugins.ckgdoku.dwp_save_err + data); 
                      }   
                    },
                'html'
            );     
 }

 /* gets both size and filetime: "size||filetime" */
 function ckged_get_unlink_size(id) {
                var params = 'call=cked_deletedsize';    params += "&cked_delid=" + id;
                jQuery.post( DOKU_BASE + 'lib/exe/ajax.php', params,   
                function (data) {  
                    if(data) { 
                     JSINFO['ckg_del_sz'] = data;
                      //console.log(data);
                    }
                      else  {
                    //      alert(LANG.plugins.ckgedit.dwp_save_err + data); 
                      }   
                    },
                'html'
            );    
       
 }
 
 function ckged_setmedia(id,del, refresh_cb) {
	 
             var params = 'call=cked_upload';    params += "&ckedupl_id=" + id;
             if(del)  params += "&ckedupl_del=D&delsize="+JSINFO['ckg_del_sz'];
                jQuery.post( DOKU_BASE + 'lib/exe/ajax.php', params,   
                function (data) {  
                    if(data) { 
                      if(refresh_cb) {
                           refresh_cb.postMessage(JSINFO['doku_url'], JSINFO['doku_url']);
                      }
                      console.log(data);
                    }
                      else  {
                    //      alert(LANG.plugins.ckgedit.dwp_save_err + data); 
                      }   
                    },
                'html'
            );    
 }
 jQuery(document).ready(function() {
     if(JSINFO['hide_captcha_error'] =='hide') {
         jQuery("div.error").hide();
     }   
 });  


jQuery(document).ready(function(){

    jQuery( "#editor_height" ).keydown(function(event) {     
          if ( event.which == 13 ) {
           event.preventDefault();
        }
    });

    $dokuWiki = jQuery('.dokuwiki');
     jQuery('.editbutton_table button').click(function() {
           var f = this.form;
           jQuery('<input />').attr('type','hidden').attr('name','mode').attr('value','dwiki').appendTo(jQuery(f));
            jQuery('<input />').attr('type','hidden').attr('name','fck_preview_mode').attr('value','nil').appendTo(jQuery(f));
    });  
  
    if(typeof(JSINFO['dbl_click_auth'] !== 'undefined') && JSINFO['dbl_click_auth'] == "") return;  
    if(!JSINFO['ckg_dbl_click']) return;
   
    /**
     * If one or more edit section buttons exist?
     * This makes sure this feature is enabled only on the edit page and for users with page edit rights.
     */
    if (jQuery('.editbutton_section', $dokuWiki).length > 0) {

        // register double click event for all headings and section divs
        jQuery('[class^="sectionedit"], div[class^="level"]', $dokuWiki).dblclick(function(){
            // find the closest edit button form to the element double clicked (downwards) and submit the form
            var f =  jQuery(this).nextAll('.editbutton_section:eq(0)').children('form:eq(0)');
            //alert(jQuery(f).hasClass('button'));            
            jQuery('<input />').attr('type','hidden').attr('name','mode').attr('value','dwiki').appendTo(jQuery(f));
            jQuery('<input />').attr('type','hidden').attr('name','fck_preview_mode').attr('value','nil').appendTo(jQuery(f));
            f.submit();
        })
    }

});
 
function ckg_edit_mediaman_insert(edid, id, opts, dw_align) {
    var link, width, s, align;

    //parse option string
    var options = opts.substring(1).split('&');

    //get width and link options
    link = 'detail';
    for (var i in options) {
        var opt = options[i];
         if (opt.match(/^\d+$/)) {   
            width = opt;
        } else  if (opt.match(/^\w+$/)) {   
            link = opt;
        }
    }

    //get alignment option
    switch (dw_align) {
    case '2':
        align = 'medialeft';
        break;
    case '3':
        align = 'mediacenter';
        break;
    case '4':
        align = 'mediaright';
        break;
    default:
        align = '';
        break;
    }

    var funcNum = CKEDITOR.instances.wiki__text._.filebrowserFn;
    var fileUrl = DOKU_BASE + 'lib/exe/fetch.php?media=' + id;
    CKEDITOR.tools.callFunction(funcNum, fileUrl, function() {
        var dialog = this.getDialog();
        if ( dialog.getName() == "image" ) {
            if (align != null) {
                dialog.getContentElement("info", "cmbAlign").setValue(align);
            }
            if (link != null) {
                dialog.getContentElement("info", "cmbLinkType").setValue(link);
            }
            if (width != null) {
                dialog.getContentElement("info", "txtWidth").setValue(width);
                dialog.dontResetSize = true;
            }
        }
    });
}

function ckg_edit_mediaman_insertlink(edid, id, opts, dw_align) {
    var funcNum = CKEDITOR.instances.wiki__text._.filebrowserFn;
    CKEDITOR.tools.callFunction(funcNum, id, function() {
        var dialog = this.getDialog();
        if (dialog.getName() == "link") {
            dialog.getContentElement('info', 'media').setValue(id);
        }
    });
}

function getCookie(name) {
    var re = new RegExp(name + "=([^;]+)");
    var value = re.exec(document.cookie);
    return (value != null) ? unescape(value[1]) : null;
}
